receiver/jaegerreceiver/factory.go

```diff
@@ -86,13 +85,39 @@ func (f *Factory) CustomUnmarshaler() component.CustomUnmarshaler {
 			return fmt.Errorf("must specify at least one protocol when using the Jaeger receiver")
 		}
 		for k := range protocols {
-			if _, ok := receiverCfg.Protocols[k]; !ok {
-				if receiverCfg.Protocols[k], err = defaultsForProtocol(k); err != nil {
-					return err
+			switch k {
+			case protoGRPC:
+				if receiverCfg.GRPC != nil {
+					continue
+				}
+				receiverCfg.GRPC = &configgrpc.GRPCServerSettings{
+					Endpoint: defaultGRPCBindEndpoint,
+				}
```

Comment:
Since you've simplified the structure of the config, I think you can simplify this code too. You don't really need anything from `GetStringMap` onwards (line 81-90).

Instead just:
```go
if receiverCfg.GRPC == nil {
	receiverCfg.GRPC = &configgrpc.GRPCServerSettings{
		Endpoint: defaultGRPCBindEndpoint,
	}
}

if receiverCfg.ThriftHTTP == nil {
	...
```

Reply:
Actually on second thoughts you can probably delete this entire function. The default behaviour in `LoadReceiver` is to create a default config and unmarshal into that. I believe that retains the default values for anything you don't explicitly override.

Reply:
... the behavior I want is that if user does not specify a protocol will be disabled, but if user specify just the protocol (no other config for that) to use defaults. The problem is that Viper does not create empty objects for when just the protocol is defined with no other config. So what I end up doing is I create the entire default config and for everything missing in Viper I remove the default config.

Reply:
Ugg yea okay seems like it will be a little bit ugly no matter which way you do it then. Now that I understand what you want to achieve I feel like maybe it was better how it was before 😆. It does feel a little odd that this way where the default config is added and then gets removed it if we later find the key wasn't there.

Reply:
I will keep this way to not have to deal with merging. This way user's config overwrites defaults.